### PR TITLE
GEOMESA-1442,GEOMESA-997,GEOMESA-535 CLI distributed export for HBase, Accumulo

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Query plans are fully serializable and used to configure m/r inputs
* Libjars files are separated out to ingest/export and back-end specific
* Hadoop serialization no longer stores the sft with each feature
* Export supports transform functions
* GML export supports v2 and v3

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by elahrvivaz** [comment]
I'll need to update this once https://github.com/locationtech/geomesa/pull/2271 is merged

File: docs/user/cli/export.rst [hunk 1]
```diff
@@ -66,29 +66,38 @@ Argument                 Description
...
 
     --attributes id,name,age,geom
 
-Transforms can be accomplished by specifying transform functions::
+Transforms can be accomplished by using ``--attribute`` to specify transform functions. Note that compared to
+``--attributes``, only a single argument can be specified per parameter, and commas will not be interpreted as
+a delimiter::
 
-    --attributes id,name,name_transform=strConcat(name, 'foo')
+    --attribute id --attribute name --attribute "name_transform=strConcat(name, 'foo')"
```
Comment (jnh5y) [inline docs/user/cli/export.rst]: Can you mix --attributes and --attribute?

File: docs/user/cli/export.rst [hunk 2]
```diff
@@ -111,8 +122,11 @@ The ``--output-format`` argument defines the encoding used for export. Currently
     map generation it is highly recommended to use GeoServer. Additionally, the resulting file from this command
     requires the use of an online browser to open in order to access online resources.
 
-The ``--output`` argument can be used to export to a file. By default, export data is written to the standard
-output stream.
+The ``--output`` argument can be used to export to a file. By default, export data is written to standard output.
+
+The ``--sort-by`` argument can be used to sort the output by one or more attributes. Note that for local exports,
+this will usually be done in-memory, so can be costly for large result sets. By default, output is sorted in
```
Comment (jnh5y) [inline docs/user/cli/export.rst]: Are there options here to spill to disk for the sort?

File: docs/user/cli/export.rst [hunk 3]
```diff
@@ -111,8 +122,11 @@ The ``--output-format`` argument defines the encoding used for export. Currently
     map generation it is highly recommended to use GeoServer. Additionally, the resulting file from this command
     requires the use of an online browser to open in order to access online resources.
 
-The ``--output`` argument can be used to export to a file. By default, export data is written to the standard
-output stream.
+The ``--output`` argument can be used to export to a file. By default, export data is written to standard output.
+
+The ``--sort-by`` argument can be used to sort the output by one or more attributes. Note that for local exports,
```
Comment (jnh5y) [inline docs/user/cli/export.rst]: Is the implication that sorting for distributed exports work well?:)

File: docs/user/cli/geotools.rst [hunk 4]
```diff
@@ -82,8 +82,8 @@ folder. To install JARs, use the script provided with the distribution:
     $ bin/install-hadoop.sh lib
 
 If you installed JARs for any additional data stores, you will need to add them to the Hadoop libjars path
-by modifying the file ``org/locationtech/geomesa/geotools/tools/ingest-libjars.list`` inside the
-``lib/geomesa-gt-tools_2.11-$VERSION.jar`` JAR.
+by modifying the file ``org/locationtech/geomesa/geotools/tools/gt-libjars.list`` inside the JAR
```
Comment (jnh5y) [inline docs/user/cli/geotools.rst]: (Idle question) Why did this filename change?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala [hunk 5]
```diff
@@ -145,49 +168,55 @@ object AccumuloQueryPlan extends LazyLogging {
...
   // join on multiple tables - requires multiple scans
-  case class JoinPlan(filter: FilterStrategy,
-                      tables: Seq[String],
-                      ranges: Seq[org.apache.accumulo.core.data.Range],
-                      iterators: Seq[IteratorSetting],
-                      columnFamily: Option[Text],
-                      numThreads: Int,
-                      joinFunction: JoinFunction,
-                      joinQuery: BatchScanPlan) extends AccumuloQueryPlan {
+  case class JoinPlan(
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala]: Do JoinPlans work in the distributed export?!?!

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AccumuloJoinIndex.scala [hunk 6]
```diff
@@ -145,7 +155,7 @@ trait AccumuloJoinIndex extends GeoMesaFeatureIndex[AttributeIndexValues[Any], A
         // finally apply the arrow iterator on the resulting features
         val (iter, reduce) = ArrowIterator.configure(transformSft, this, ds.stats, None, None, hints)
         val iters = visibilityIter(indexSft) ++ Seq(filterTransformIter, rowValueIter, iter)
-        plan(iters, ArrowIterator.kvsToFeatures(), Some(reduce))
+        plan(iters, new AccumuloArrowResultsToFeatures(), Some(reduce))
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AccumuloJoinIndex.scala]: (Curiosity) Why the change from the object.def to having classes here?  (I'm guessing it is related to making plans serializable?)

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/ArrowIterator.scala [hunk 7]
```diff
@@ -56,15 +55,9 @@ object ArrowIterator {
...
     */
-  def kvsToFeatures(): Entry[Key, Value] => SimpleFeature = {
-    val sf = new ScalaSimpleFeature(ArrowEncodedSft, "")
-    sf.setAttribute(1, GeometryUtils.zeroPoint)
-    e: Entry[Key, Value] => {
-      sf.setAttribute(0, e.getValue.get())
-      sf
-    }
+  class AccumuloArrowResultsToFeatures extends ArrowResultsToFeatures[Entry[Key, Value]] {
+    override protected def bytes(result: Entry[Key, Value]): Array[Byte] = result.getValue.get()
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/ArrowIterator.scala]: Oh!  Having the classes lets one centralize the above code and be a little more DRY?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/BinAggregatingIterator.scala [hunk 8]
```diff
@@ -114,8 +103,15 @@ object BinAggregatingIterator extends LazyLogging {
...
+        new ScalaSimpleFeature(BinEncodedSft, deserialized.getID, values)
       }
     }
   }
+
+  /**
+    * Adapts the iterator to create simple features
+    */
+  class AccumuloBinResultsToFeatures extends BinResultsToFeatures[Entry[Key, Value]] {
+    override protected def bytes(result: Entry[Key, Value]): Array[Byte] = result.getValue.get()
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/BinAggregatingIterator.scala]: As an over-refactoring, it seems like you'd be able to have a trait for Accumulo to mix-in this def;).

File: geomesa-accumulo/geomesa-accumulo-jobs/src/main/scala/org/locationtech/geomesa/jobs/mapreduce/GeoMesaAccumuloInputFormat.scala [hunk 9]
```diff
@@ -11,124 +11,191 @@ package org.locationtech.geomesa.jobs.mapreduce
...
+    * Our delegated AccumuloInputFormat creates a split for each range - because we set a lot of ranges in
+    * geomesa, that creates too many mappers. Instead, we try to group the ranges by tservers. We use the
+    * location assignment of the tablets to tservers to determine the number of splits returned.
+    */
+  override def getSplits(context: JobContext): java.util.List[InputSplit] = {
+    val accumuloSplits = delegate.getSplits(context)
+    // Get the appropriate number of mapper splits using the following priority
+    // 1. Get splits from AccumuloMapperProperties.DESIRED_ABSOLUTE_SPLITS (geomesa.mapreduce.splits.max)
+    // 2. Get splits from #tserver locations * AccumuloMapperProperties.DESIRED_SPLITS_PER_TSERVER (geomesa.mapreduce.splits.tserver.max)
+    // 3. Get splits from AccumuloInputFormat.getSplits(context)
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-jobs/src/main/scala/org/locationtech/geomesa/jobs/mapreduce/GeoMesaAccumuloInputFormat.scala]: Is this priority listed in the docs?

File: geomesa-accumulo/geomesa-accumulo-tools/src/main/resources/org/locationtech/geomesa/accumulo/tools/ingest-libjars.list [hunk 10]
```diff
@@ -1,46 +0,0 @@
-accumulo-core
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-tools/src/main/resources/org/locationtech/geomesa/accumulo/tools/ingest-libjars.list]: How is the classpath for distributed ingest handled now?

File: geomesa-accumulo/geomesa-accumulo-tools/src/test/scala/org/locationtech/geomesa/accumulo/tools/export/AccumuloExportCommandTest.scala [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        val file = s"$out/${format.name}/project/out.${format.extensions.head}"
+        val command = connectedCommand()
+        command.params.file = file
+        command.params.attributes = List("dtg", "geom", "id").asJava
+        command.execute()
+        val tsft = SimpleFeatureTypes.createType(sft.getTypeName, "dtg:Date,*geom:Point:srid=4326")
+        readFeatures(format, file, tsft) must containTheSameElementsAs(features.map(ScalaSimpleFeature.retype(tsft, _)))
+      }
+    }
+    "support sorting" in {
```
Comment (jnh5y) [inline geomesa-accumulo/geomesa-accumulo-tools/src/test/scala/org/locationtech/geomesa/accumulo/tools/export/AccumuloExportCommandTest.scala]: Should we test both sort orders?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/aggregators/HBaseArrowAggregator.scala [hunk 12]
```diff
@@ -43,8 +38,12 @@ object HBaseArrowAggregator {
...
                 filter: Option[Filter],
                 ecql: Option[Filter],
-                hints: Hints): (Map[String, String], QueryPlan.Reducer) = {
+                hints: Hints): (Map[String, String], FeatureReducer) = {
     val conf = ArrowScan.configure(sft, index, stats, filter, ecql, hints)
     (conf.config + (GeoMesaCoprocessor.AggregatorClass -> classOf[HBaseArrowAggregator].getName), conf.reduce)
   }
+
+  class HBaseArrowResultsToFeatures extends ArrowResultsToFeatures[Array[Byte]] {
+    override protected def bytes(result: Array[Byte]): Array[Byte] = result
```
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/aggregators/HBaseArrowAggregator.scala]: Some over-refactoring idea:)  Maybe there could be a trait for this?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/package.scala [hunk 13]
```diff
@@ -325,4 +328,54 @@ package object api {
...
+  object SerializableState {
+
+    /**
+      * Serialize an object
+      *
+      * @param obj object to serialize
+      * @return
+      */
+    def serialize(obj: SerializableState): String =
+      s"${obj.getClass.getName},${StringSerialization.encodeMap(obj.state)}"
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/package.scala]: Sanity check:  Does the map have more commas in it?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/Reprojection.scala [hunk 14]
```diff
@@ -16,43 +16,90 @@ import org.locationtech.geomesa.features.ScalaSimpleFeature
...
+    *
+    * @param returnSft simple feature type being returned
+    * @param crs crs information from a query
+    * @return
+    */
+  def apply(returnSft: SimpleFeatureType, crs: QueryReferenceSystems): Reprojection = {
+    if (crs.target != crs.user) {
+      val transformer = new GeometryCoordinateSequenceTransformer
+      transformer.setMathTransform(CRS.findMathTransform(crs.user, crs.target, true))
+      val transformed = FeatureTypes.transform(returnSft, crs.target) // note: drops user data
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/Reprojection.scala]: Which userData is dropped?

File: docs/user/cli/export.rst (hunk 2 above)
Comment (elahrvivaz) [reply docs/user/cli/export.rst]: no, we haven't implemented that AFAIK, it just sets the query sort and the data store handles it as normal.

File: docs/user/cli/export.rst (hunk 3 above)
Comment (elahrvivaz) [reply docs/user/cli/export.rst]: I'm not sure what you mean by 'well', but it works :)

File: docs/user/cli/geotools.rst (hunk 4 above)
Comment (elahrvivaz) [reply docs/user/cli/geotools.rst]: I separated out the base jars used by ingest into the tools jar, then have a back-end-specific file for each distro (as noted in the commit message :).

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala (hunk 5 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala]: No, it used the `AccumuloJobUtils.getSingleQueryPlan` method, which will disable join scans

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AccumuloJoinIndex.scala (hunk 6 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AccumuloJoinIndex.scala]: Yeah, so that they can be serialized

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/ArrowIterator.scala (hunk 7 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/ArrowIterator.scala]: To some extent yeah, but the main driver was being able to serialize them.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/BinAggregatingIterator.scala (hunk 8 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/BinAggregatingIterator.scala]: haha, seems like overkill for one line :)

File: geomesa-accumulo/geomesa-accumulo-jobs/src/main/scala/org/locationtech/geomesa/jobs/mapreduce/GeoMesaAccumuloInputFormat.scala (hunk 9 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-jobs/src/main/scala/org/locationtech/geomesa/jobs/mapreduce/GeoMesaAccumuloInputFormat.scala]: I think so here: https://www.geomesa.org/documentation/user/accumulo/configuration.html#map-reduce-input-splits-properties
This didn't change, I just moved the code around a bit.

File: geomesa-accumulo/geomesa-accumulo-tools/src/main/resources/org/locationtech/geomesa/accumulo/tools/ingest-libjars.list (hunk 10 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-tools/src/main/resources/org/locationtech/geomesa/accumulo/tools/ingest-libjars.list]: The main file is [here](https://github.com/elahrvivaz/geomesa/blob/fcr_distributed_export/geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/export/ExportCommand.scala#L55) and then each back-end has a supplemental file aka [here](https://github.com/elahrvivaz/geomesa/blob/fcr_distributed_export/geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/AccumuloDataStoreCommand.scala#L61-L62) and [here](https://github.com/elahrvivaz/geomesa/blob/fcr_distributed_export/geomesa-accumulo/geomesa-accumulo-tools/src/main/scala/org/locationtech/geomesa/accumulo/tools/export/AccumuloExportCommand.scala#L24)

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/aggregators/HBaseArrowAggregator.scala (hunk 12 above)
Comment (elahrvivaz) [reply geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/aggregators/HBaseArrowAggregator.scala]: It's so trivial, I'd lean towards not...

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/package.scala (hunk 13 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/api/package.scala]: It does, but we only split on the first comma to get the class name, then pass the rest back to `StringSerialization.decodeMap` which handles them appropriately.

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/Reprojection.scala (hunk 14 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/Reprojection.scala]: it doesn't copy the sft user data, hence we add it back in the next line.

File: geomesa-accumulo/geomesa-accumulo-tools/src/test/scala/org/locationtech/geomesa/accumulo/tools/export/AccumuloExportCommandTest.scala (hunk 11 above)
Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-tools/src/test/scala/org/locationtech/geomesa/accumulo/tools/export/AccumuloExportCommandTest.scala]: added

File: docs/user/cli/geotools.rst (hunk 4 above)
Comment (jnh5y) [reply docs/user/cli/geotools.rst]: Oh man, I read +7k lines and -5k lines and I miss the commit message.

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala (hunk 5 above)
Comment (jnh5y) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala]: Is that documented?  (Do we care?)

Comment (elahrvivaz) [reply geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloQueryPlan.scala]: ah, not sure - it prints out some warnings i believe.
